### ISSUE TITLE
NS2001: Is an accessible constructor strictly necessary?

### ISSUE THREAD

**Issue by stogle** [nsubstitute/NSubstitute.Analyzers#227]
In my unit test I am creating a substitute for [this](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Media/DrawingContext.cs) third-party public abstract class, which has only an empty internal constructor. I am getting an [NS2001](https://github.com/nsubstitute/NSubstitute.Analyzers/blob/dev/documentation/rules/NS2001.md) warning because the constructor is not accessible from the test, however the substitute and my test appear to work as intended. All my test does with the substitute is pass it to my method-under-test, then manually check its `ReceivedCalls()` list. It does not set any return values or anything else. (I can't share my actual test for IP reasons, but I can create a minimal example if that would help.)

I understand the [problems](https://nsubstitute.github.io/help/creating-a-substitute/#substituting-infrequently-and-carefully-for-classes) associated with substituting for classes rather than interfaces, and that non-virtual code is executed during my test. What I'm not clear on is the specific problems with violating NS2001 by substituting for a class with an inaccessible constructor. The rule says that to fix this violation I should add an accessible constructor to the type, but I can't do that because I don't own the type. Furthermore, NSubstitute is still able to create a substitute without an accessible constructor (presumably via reflection).

Could the documentation for the rule expand on why an accessible constructor is important and what happens when you create a substitute for a class without one?

**Comment by tpodolak** [comment]
Hi @stogle 
>  Is an accessible constructor strictly necessary

Yes it is. Here is a minimal repro showing that NSubstitute will fail to create substitute for class with only internal ctor

```c#
public class Test
{
    [Fact]
    public void Foo()
    {
        var substitute = Substitute.For<MyAbstractClass>();
        _ = substitute;
    }
}

public abstract class MyAbstractClass
{
    internal MyAbstractClass()
    {
    }
}
```

![image](https://github.com/user-attachments/assets/88cc93c2-7e8b-4b3e-bbf1-da15f8ca1419)


When it comes to your case, it looks like Avalonia marks their assembly with `[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")` [attribute](https://github.com/AvaloniaUI/Avalonia/blob/aabd038e88951a992124ddfcb2be06ee6daaf4a5/src/Avalonia.Base/Properties/AssemblyInfo.cs#L36) basically making iternal classes and constructors visible to Castle.Core(which is used by NSubstitute to create proxy classes) - thats why NSubstitute is able to instantiate a substitute. However analyzers were not able to find this annotation(even though we check for that attribute being applied to [assembly](https://github.com/nsubstitute/NSubstitute.Analyzers/blob/a3704ba71b247e8d39533d83094bb15fd556f8d5/src/NSubstitute.Analyzers.Shared/Extensions/ISymbolExtensions.cs#L13)) - thats why you see the warning. I will take a look if thats a bug on our end, or if for whatever reasons Roslyn is not able to find these annotations

**Comment by stogle** [comment]
Hi @tpodolak, thanks for the answer! I did not think to look for the InternalsVisibleTo attribute, but that makes total sense.

> I will take a look if thats a bug on our end, or if for whatever reasons Roslyn is not able to find these annotations

Either way, I now have a proper justification for disabling the warning in this case. Thanks again.

**Comment by tpodolak** [comment]
Just did a quick check, and Roslyn doesnt return any constructors for this particular type. However this doesnt seem to be Roslyn bug, but rather some sort of compilation option of Avalonia. Internals members visibility seems to be disabled
![image](https://github.com/user-attachments/assets/42999039-f479-4ca9-a594-f865c4a994b4)
so even though ctor can be found in the metadata
![image](https://github.com/user-attachments/assets/ff5a63c8-4d46-4e6c-b52e-8652262ac81c)
it will not surface as part of `GetMembers` call as
```c#
if (isOrdinaryEmbeddableStruct || module.ShouldImportMethod(methodHandle, moduleSymbol.ImportOptions))
```
returns false

**Comment by stogle** [comment]
I'm not that familiar with Roslyn or writing analyzers, but I think what you are saying is that NSubstitute can see this constructor in Avalonia.Base.dll at runtime, but NSubstitute.Analyzers can't see it during analysis, even though it normally sees internal constructors on other assemblies that use `InternalsVisibleTo`. Is that it in a nutshell?

If it would help, I could create an Issue or a Discussion in the Avalonia repository about this. I'm just not sure what to say.

Still, using a #pragma to disable this particular warning isn't a big deal for me, as long as I have a justification.
